lib/DiscoveryStreamFeed.jsm

```diff
@@ -240,8 +240,16 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
         // we then fill in with the proper object inside the promise.
         newFeeds[url] = {};
         const feedPromise = this.getComponentFeed(url, isStartup);
+
         feedPromise.then(feed => {
           newFeeds[url] = this.filterRecommendations(feed);
+          sendUpdate({
+            type: at.DISCOVERY_STREAM_FEED_UPDATE,
+            data: {
+              feed: newFeeds[url],
+              url,
+            },
+          });
```

Comment:
That's where we are hoping big win by updating state right when we are ready with feed data. It's sequential that we are invoking buildFeedPromise in sequential manner, though its possible for a higher (lower on the page) feedPromise to resolve early and sendUpdate , is that correct?

Reply:
> That's where we are hoping big win by updating state right when we are ready with feed data. It's sequential that we are invoking buildFeedPromise in sequential manner, though its possible for a higher (lower on the page) feedPromise to resolve early and sendUpdate , is that correct?
Tests below "should stop rendering feeds if we hit one that's not ready" help explain component will be rendered on the page sequentially

Reply:
Yup, it can be ready sooner, and when we go to render, we wouldn't attempt to render it unless everything before it is also ready.

So think of the rendering as sequential and not the fetching.

I tested a version where the fetches were also sequential, where the `feedPromise.then(feed => {` would chain, but it was 1. Way more complex, 2. Didn't impress me as being much faster, and 3. Could actually be slower.

The way I have it, if the last feed to load is blocking others, once it's done, instead of now having to fire off a bunch more `DISCOVERY_STREAM_FEED_UPDATE` we can just fire off the last 1 and the rest of the data should just be there and things should render completely in one go.